src/strands/multiagent/function_node.py

```diff
@@ -0,0 +1,157 @@
+"""FunctionNode implementation for executing deterministic Python functions as graph nodes.
+
+This module provides the FunctionNode class that extends MultiAgentBase to execute
+regular Python functions while maintaining compatibility with the existing graph
+execution framework, proper error handling, metrics collection, and result formatting.
+"""
+
+import logging
+import time
+from typing import Any, Protocol, Union
+
+from opentelemetry import trace as trace_api
+
+from ..agent import AgentResult
+from ..telemetry import get_tracer
+from ..telemetry.metrics import EventLoopMetrics
+from ..types.content import ContentBlock, Message
+from ..types.event_loop import Metrics, Usage
+from .base import MultiAgentBase, MultiAgentResult, NodeResult, Status
+
+logger = logging.getLogger(__name__)
+
+
+class FunctionNodeCallable(Protocol):
+    """Protocol for functions that can be executed within FunctionNode."""
+
+    def __call__(
+        self, task: str | list[ContentBlock], invocation_state: dict[str, Any] | None = None, **kwargs: Any
+    ) -> str | list[ContentBlock] | Message:
```

Comment:
is there any way we can provide more flexibility here? This essentially gets them to write MultiAgentBase. I'd rather they return string or something serializable, and we adjust accordingly